addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java

```diff
@@ -0,0 +1,237 @@
+ *
+ */
+@Component(service = MDNSDiscoveryParticipant.class, immediate = true)
+public class DenonMarantzDiscoveryParticipant implements MDNSDiscoveryParticipant {
+
+    private Logger logger = LoggerFactory.getLogger(DenonMarantzDiscoveryParticipant.class);
+
+    // Service type for 'Airplay enabled' receivers
+    private static final String RAOP_SERVICE_TYPE = "_raop._tcp.local.";
+
+    /**
+     * Match the serial number, vendor and model of the discovered AVR.
+     * Input is like "0006781D58B1@Marantz SR5008._raop._tcp.local."
+     * A Denon AVR serial (MAC address) starts with 0005CD
+     * A Marantz AVR serial (MAC address) starts with 000678
+     */
+    private static final Pattern DENON_MARANTZ_PATTERN = Pattern
+            .compile("^((?:0005CD|000678)[A-Z0-9]+)@(.+)\\._raop\\._tcp\\.local\\.$");
+
+    /**
+     * Denon AVRs have a MAC address / serial number starting with 0005CD
+     */
+    private static final String DENON_MAC_PREFIX = "0005CD";
+
+    /**
+     * Marantz AVRs have a MAC address / serial number starting with 000678
+     */
+    private static final String MARANTZ_MAC_PREFIX = "000678";
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return Collections.singleton(THING_TYPE_AVR);
+    }
+
+    @Override
+    public String getServiceType() {
+        return RAOP_SERVICE_TYPE;
+    }
+
+    @Override
+    public DiscoveryResult createResult(ServiceInfo serviceInfo) {
+        String qualifiedName = serviceInfo.getQualifiedName();
+        logger.debug("AVR found: {}", qualifiedName);
+        ThingUID uid = getThingUID(serviceInfo);
+        if (uid != null) {
+            Matcher matcher = DENON_MARANTZ_PATTERN.matcher(qualifiedName);
+            matcher.matches(); // we already know it matches, it was matched in getThingUID
+            String serial = matcher.group(1).toLowerCase();
+
+            /**
+             * The Vendor is not available from the mDNS result.
+             * We assign the Vendor based on our assumptions of the MAC address prefix.
+             */
+            String vendor = "";
+            if (serial.startsWith(MARANTZ_MAC_PREFIX)) {
+                vendor = "Marantz";
+            } else if (serial.startsWith(DENON_MAC_PREFIX)) {
+                vendor = "Denon";
+            }
+
+            // 'am=...' property describes the model name
+            String model = serviceInfo.getPropertyString("am");
+            String friendlyName = matcher.group(2).trim();
+
+            Map<String, Object> properties = new HashMap<>(2);
+
+            if (serviceInfo.getHostAddresses().length == 0) {
+                logger.debug("Could not determine IP address for the Denon/Marantz AVR");
+                return null;
+            }
+            String host = serviceInfo.getHostAddresses()[0];
+
+            logger.debug("IP Address: {}", host);
+
+            // try a HTTP request to autoconfigure the HTTP / Telnet parameter
```

Comment:
I agree it's not perfect. For some models HTTP is not an option, so for those I want to set 'Use Telnet' to `true` in the Thing parameter. The default for this parameter is `false`.

Should I do this instead?
- set some property in the Thing indicating it was created via a Discovery Result (I guess there is no way to determine otherwise?)
- if set, auto-detect and alter the value for parameter 'Use Telnet' during Inititialization

E.g. I don't want to override the parameter when "explicitly" (there is a default) set by the user during manual Thing creation (for whatever reason).

Reply:
I found another option, but have trouble implementing this:
Limit the number of times the procedure is invoked to exactly once, by stopping further (relatively expensive) discovery activities as soon as it's clear we already configured or discovered the Thing.
https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html#compare-with-existing-results

I tried to implement the `ExtendedDiscoveryService` with the `MDNSDiscoveryParticipant`, this does not work unfortunately. How can I add this callback to the MDNSDiscoveryService? How do I approach this?

Reply:
@jwveldhuis This is a missing feature which I have just implemented in #5526

Reply:
@triller-telekom perfect!

Reply:
I am not sure whether this would be a solution. After all, you still have to do HTTP requests in a method that must return fast. And discovery is supposed to be passive in most cases, it should really just sit there and listen.

Couldn't you simply leave this property unset (=null) and upon handler initialisation, you do the HTTP call and figure out the right value, if it is still null? This would allow you to set the value without bothering the user and still avoid the risk of overwriting a value that the user has specifically set.

Reply:
Ok, I have now moved this part over the the Handler. Added benefit is the auto configuration for telnet/http and zone count is now also applied to Things configured in text files.